StRoot/StFcsWaveformFitMaker/StFcsWaveformFitMaker.cxx

```diff
@@ -216,6 +379,68 @@ int StFcsWaveformFitMaker::Finish(){
     c->cd(3)->SetLogy(); mTimeIntg[2]->Draw("colz");
     c->cd(4)->SetLogy(); mTimeIntg[3]->Draw("colz");
     c->SaveAs("stage0.png");
+    delete c;
+  }
+  if( mOutFile!=0 ){
+    //mOutFile = new TFile(mOutFileName.c_str(),"RECREATE");
+    mOutFile->cd();
+    
+    for( UShort_t i=0; i<7; ++i ){
+      if( i<3 ){
+	if( mH2_Dep0DepMod[i]!=0 ){mH2_Dep0DepMod[i]->Write();}
+	if( mH2_Sum8Dep0[i]!=0 )  {mH2_Sum8Dep0[i]->Write();}
+	if( mH2_Sum8DepMod[i]!=0 ){mH2_Sum8DepMod[i]->Write();}
+      }
+      if( i<6 ){
+	if( mH2F_AdcTbAkio[i]!=0 )     { mH2F_AdcTbAkio[i]->Write();}
+	if( mH2F_AdcTbMine[i]!=0 )     { mH2F_AdcTbMine[i]->Write();}
+	if( mH2F_SumFitvSumWin[i]!=0 ) { mH2F_SumFitvSumWin[i]->Write();}
+	if( mH2F_APeakvMPeak[i]!=0 )   { mH2F_APeakvMPeak[i]->Write();}
+	if( mH1F_PeakStart[i]!=0 )     { mH1F_PeakStart[i]->Write();}
+	if( mH1F_PeakEnd[i]!=0 )       { mH1F_PeakEnd[i]->Write();}
+	if( mH2F_NOvsId[i]!=0    )     { mH2F_NOvsId[i]->Write();}
+      }
+      if( mH2F_AdcTbValidPeak[i]!=0 ){ mH2F_AdcTbValidPeak[i]->Write();}
+      if( mH1F_NPeaks[i]!=0 )        { mH1F_NPeaks[i]->Write();}
+      if( mH1F_NPeaksFiltered[i]!=0 ){ mH1F_NPeaksFiltered[i]->Write();}
+      
+      if( mH1F_Res0[i]!=0      )     { mH1F_Res0[i]->Write();}
+      if( mH1F_Res0Zoom[i]!=0  )     { mH1F_Res0Zoom[i]->Write();}
+      if( mH1F_Sum8Res0[i]!=0  )     { mH1F_Sum8Res0[i]->Write();}
+      if( mH1F_Sum8Res0Zoom[i]!=0 )  { mH1F_Sum8Res0Zoom[i]->Write();}
+      if( mH1F_FitRes0[i]!=0      )  { mH1F_FitRes0[i]->Write();}
+      if( mH1F_FitRes0Zoom[i]!=0  )  { mH1F_FitRes0Zoom[i]->Write();}
+      if( mH2F_Sum8vFit[i]!=0     )  { mH2F_Sum8vFit[i]->Write();}
+    }
+    
+    if( mTime!=0 ){ mTime->Write(); }
+    
+    if( mH1_NPeaksAkio!=0 ){ mH1_NPeaksAkio->Write(); }
+    if( mH1_NPeaksFilteredAkio!=0 ){ mH1_NPeaksFilteredAkio->Write(); }
+    if( mH1_PeakTiming!=0 ){mH1_PeakTiming->Write();}
+    
+    if( mH2_NPeakvsPeakXdiff!=0 )      {mH2_NPeakvsPeakXdiff->Write();}
+    if( mH2_NPeakvsPeakYratio!=0)      {mH2_NPeakvsPeakYratio->Write();}
+    if( mH1_VOverlap!=0         )      {mH1_VOverlap->Write();}
```

Comment:
Have you tried to make a single call to `rootFile->Write();` instead of doing this for individual objects?

Reply:
I have tried it before for other classes and it doesn't always work. It probably has to do with the order in which you create TFile and histograms so I explicitly write them in now. If it is really a big problem I can try to see if rootFile->Write() will work in this instance.

Reply:
I tried with just rootFile->Write() and it didn't work. I understand the issue but I don't know what else to do to avoid this. You are correct that if no test level is called then no output needs to be saved. The test levels are for single file tests and should never be called in production. They are intended to help understand and improve the algorithm rather than for generating generic QA histograms for StFcsWaveformFitMaker.